From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stanvonderhtlwienwestlol <dusek.s03@htlwienwest.at>
Date: Wed, 22 Dec 2021 15:44:09 +0100
Subject: [PATCH] Fixed the Previous Fix


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 19b04e960d0923a61ceb6ea5be4d277be07383f7..80546463ff774a23a5e73d97460edc0df7591aa6 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -762,7 +762,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
             if (AtlasConfig.mergeEntities == true) {
                 final Entity[] prev = {null};
                 this.entityTickList.forEach(entity -> {
-                    if (prev[0] instanceof PrimedTnt || prev[0] instanceof FallingBlockEntity) {
+                    if (entity instanceof PrimedTnt || entity instanceof FallingBlockEntity) {
                         if (prev[0] instanceof PrimedTnt && entity instanceof PrimedTnt && ((PrimedTnt) prev[0]).equalsMerge((PrimedTnt) entity)) {
                             ((PrimedTnt) entity).count += ((PrimedTnt) prev[0]).count;
                             prev[0].discard();
