From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: stanvonderhtlwienwestlol <dusek.s03@htlwienwest.at>
Date: Tue, 22 Feb 2022 16:27:40 +0100
Subject: [PATCH] Fixed Swing on Tunneled Hammer


diff --git a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
index bbf2de88466264f133c797115597ddf6f7e8c38c..be9f9c62a43f00825494bfd5e6eb90b7ba896037 100644
--- a/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/src/main/java/net/minecraft/world/entity/item/PrimedTnt.java
@@ -72,7 +72,25 @@ public class PrimedTnt extends Entity {
     @Override
     public void tick() {
         //jesus modification - execute merged tnt
-        PrimedTnt initial = this.clone();
+        if (this.getFuse() - 1 <= 0 && this.count > 1) {
+            PrimedTnt vtnt = this.clone();
+            vtnt.setFuse(100);
+            vtnt.tick();
+
+            if (AtlasConfig.CombinedExplosions &&
+                ((this.getDeltaMovement().x == 0 &&
+                this.getDeltaMovement().y == 0 &&
+                this.getDeltaMovement().z == 0) || this.distanceTo(vtnt) > 8)&&
+                vtnt.level.getBlockState(vtnt.blockPosition()).getMaterial().isLiquid()) {
+                this.cancombine = true;
+            } else {
+                for (int x = 0; x < this.count -1; x++) {
+                    PrimedTnt merged = this.clone();
+                    merged.tick();
+                }
+            }
+            vtnt.discard();
+        }
 
         if (level.spigotConfig.currentPrimedTnt++ > level.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
         if (!this.isNoGravity()) {
@@ -92,21 +110,6 @@ public class PrimedTnt extends Entity {
             this.setDeltaMovement(this.getDeltaMovement().multiply(0.7D, -0.5D, 0.7D));
         }
 
-        if (this.getFuse() - 1 <= 0 && this.count > 1) {
-            if (AtlasConfig.CombinedExplosions &&
-                this.getDeltaMovement().x == 0 &&
-                this.getDeltaMovement().y == 0 &&
-                this.getDeltaMovement().z == 0 &&
-                this.level.getBlockState(this.blockPosition()).getMaterial().isLiquid()) {
-                this.cancombine = true;
-            } else {
-                for (int x = 0; x < this.count -1; x++) {
-                    PrimedTnt merged = initial.clone();
-                    merged.tick();
-                }
-            }
-        }
-
         int i = this.getFuse() - 1;
 
         this.setFuse(i);
